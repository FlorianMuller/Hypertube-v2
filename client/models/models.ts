import { AxiosResponse, AxiosError, Canceler } from "axios";

export interface Locale {
  locale: string;
  setLocale: (locale: string) => void;
}

export type Method = "get" | "put" | "post" | "patch" | "delete";
export type Headers = { [key: string]: string } | undefined;
export type Data = object | undefined;

export interface UseApiOption {
  method?: Method;
  headers?: Headers;
  data?: Data;
  hotReload?: boolean;
  validateStatus?: (status: number) => boolean;
}

export interface UseApiReturn<T, E> {
  callApi: (freshData?: Data) => Promise<void>;
  loading: boolean;
  res: AxiosResponse<T>;
  resData: T;
  error: AxiosError<E>;
  setUrl: (url: string) => void;
  setMethod: (method: Method) => void;
  setHeaders: (headers: Headers) => void;
  setData: (data: Data) => void;
  cancelAllRequests: Canceler;
}

export interface ApiAuthResponse {
  validToken: boolean;
  error?: string;
}

export interface AuthInfo {
  username: string;
  password: string;
  [key: string]: string;
}

export interface Comment {
  movieName: string;
  date: number;
  name: string;
  body: string;
  stars: number;
  _id: number;
}

export interface Review {
  id: string;
  name: string;
  date: number;
  stars: number;
  body: string;
}

export interface Filters {
  query: string;
  collections: string[];
  year: number;
  minRating: number;
}

export interface Movie {
  id: string;
  title: string;
  cover: string;
  year: number;
  summary: string;
  genres: string[];
  rating: number;
  runtime: number;
}

export interface ApiSearchReponse {
  movies: Movie[];
  nextPage: boolean;
}

export type CustomSnackbarVariant =
  | "success"
  | "warning"
  | "error"
  | "info"
  | "default";
