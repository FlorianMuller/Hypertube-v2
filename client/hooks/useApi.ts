import { useEffect, useState } from "react";

import API from "../util/api";

import { Fixture } from "../models/models";

const useApi = (
  url: string,
  fixture?: Fixture
): { data: Record<string, unknown>; loading: boolean; error: {} } => {
  const [data, setData] = useState({});
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    if (fixture) {
      setData(fixture.data);
      setLoading(fixture.loading);
      setError(fixture.error);
    } else {
      setLoading(true);
      setError(null);

      API.get(url)
        .then((res) => {
          setData(res.data);
        })
        .catch((err) => {
          setError(err);
        })
        .finally(() => {
          setLoading(false);
        });
    }
  }, [fixture, url]);
  return { data, loading, error };
};

export default useApi;
