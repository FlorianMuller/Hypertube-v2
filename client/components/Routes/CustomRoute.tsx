import React, { ReactElement, ElementType } from "react";
import { Redirect, Route } from "react-router";

import useApi from "../../hooks/useApi";
import { Fixture, ApiAuthResponse } from "../../models/models";

import Loading from "./Loading";

interface Props {
  authComponent?: ElementType;
  notAuthComponent?: ElementType;
  exact?: boolean;
  path?: string;
  fixture?: Fixture;
}

const CustomRoute = ({
  authComponent: AuthComponent,
  notAuthComponent: NotAuthComponent,
  path,
  exact,
  fixture
}: Props): ReactElement => {
  const { data, loading, error } = useApi<ApiAuthResponse, ApiAuthResponse>(
    "/check-auth",
    fixture
  );

  return (
    <Route
      path={path}
      exact={exact}
      render={(): ReactElement => {
        if (error?.response?.status >= 500 && error.response.status <= 599)
          return <div>Error</div>;
        if (loading) return <Loading />;

        if (AuthComponent && data?.validToken) return <AuthComponent />;
        if (NotAuthComponent && !data?.validToken) return <NotAuthComponent />;
        if (!loading) return <Redirect to="/" />;

        return null;
      }}
    />
  );
};

export default CustomRoute;
