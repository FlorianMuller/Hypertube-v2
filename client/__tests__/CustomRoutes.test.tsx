import React, { ReactElement } from "react";
import renderer from "react-test-renderer";
import { Route, MemoryRouter } from "react-router-dom";

import CustomRoute from "../components/Routes/CustomRoute";

import useApi from "../hooks/useApi";
import { UseApiReturn, ApiAuthResponse } from "../models/models";

jest.mock("../hooks/useApi", () => jest.fn());

const mockUseApi = useApi as jest.Mock<
  UseApiReturn<ApiAuthResponse, ApiAuthResponse>
>;

const basicUseApiValue: UseApiReturn<ApiAuthResponse, ApiAuthResponse> = {
  callApi: jest.fn(),
  loading: false,
  res: null,
  resData: null,
  error: null,
  setUrl: jest.fn(),
  setMethod: jest.fn(),
  setHeaders: jest.fn(),
  setData: jest.fn(),
  cancelAllRequests: jest.fn()
};

interface Props {
  children: ReactElement;
}

const MockRouter = ({ children }: Props): ReactElement => {
  return (
    <MemoryRouter initialEntries={["/test"]}>
      {children}

      <Route exact path="/">
        <div>Home/Login</div>
      </Route>
    </MemoryRouter>
  );
};

describe("CustomRoute", () => {
  it("Is connected, private route : should show div", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      resData: { validToken: true }
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          authComponent={(): ReactElement => <div>Should be shown</div>}
        />
      </MockRouter>
    );

    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });

  it("Is connected, public route : should redirect to Home/Login", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      resData: { validToken: true }
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          notAuthComponent={(): ReactElement => <div>Should NOT be shown</div>}
        />
      </MockRouter>
    );

    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });

  it("Is not connected, private route : should redirect to Home/Login", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      resData: { validToken: false }
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          authComponent={(): ReactElement => <div>Should NOT be shown</div>}
        />
      </MockRouter>
    );

    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });

  it("Is not connected, public route : should show div", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      resData: { validToken: false }
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          notAuthComponent={(): ReactElement => <div>Should be shown</div>}
        />
      </MockRouter>
    );

    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });

  it("Is connected, public/private route : should show private", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      resData: { validToken: true }
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          notAuthComponent={(): ReactElement => <div>Public component</div>}
          authComponent={(): ReactElement => <div>Private component</div>}
        />
      </MockRouter>
    );

    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });

  it("Is not connected, public/private route : should show public", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      resData: { validToken: false }
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          notAuthComponent={(): ReactElement => <div>Public component</div>}
          authComponent={(): ReactElement => <div>Private component</div>}
        />
      </MockRouter>
    );

    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });

  it("Loading", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      loading: true
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          notAuthComponent={(): ReactElement => <div>Loading</div>}
        />
      </MockRouter>
    );
    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });

  it("Error", () => {
    mockUseApi.mockImplementation(() => ({
      ...basicUseApiValue,
      error: {
        response: {
          data: { validToken: false },
          status: 500,
          statusText: "",
          headers: null,
          config: null
        },
        config: null,
        name: null,
        message: null,
        isAxiosError: true,
        toJSON: null
      }
    }));
    const domNode = (
      <MockRouter>
        <CustomRoute
          path="/test"
          exact
          notAuthComponent={(): ReactElement => <div>Should NOT be shown</div>}
        />
      </MockRouter>
    );

    const tree = renderer.create(domNode);
    tree.update(domNode);
    expect(tree).toMatchSnapshot();
  });
});
