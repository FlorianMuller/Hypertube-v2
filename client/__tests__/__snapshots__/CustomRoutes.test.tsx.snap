// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomRoute Error 1`] = `
<div>
  Error
</div>
`;

exports[`CustomRoute Is connected, private route : should show div 1`] = `
<div>
  Should be shown
</div>
`;

exports[`CustomRoute Is connected, public route : should NOT show div 1`] = `null`;

exports[`CustomRoute Is not connected, private route : should NOT show div 1`] = `null`;

exports[`CustomRoute Is not connected, public route : should show div 1`] = `
<div>
  Should be shown
</div>
`;

exports[`CustomRoute Loading 1`] = `
<div
  className="MuiGrid-root makeStyles-center-1 MuiGrid-container"
>
  <div
    className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;
